### PR TITLE
coreboot support

### PR DESCRIPTION
**Pull request by retrage**
This PR proposes "coreboot" feature for rust-hypervisor-firmware as a coreboot payload.
It adds `coreboot::StartInfo` that parses coreboot table[1] to provide e820 table and RSDP address. (This part may be able to reuse for PR #24)
It adds `rust64_start` for "coreboot" feature because the `coreboot::StartInfo` struct should be allocated and supplied before `main`.

Since coreboot payload starts from 32-bit protected mode, and it uses the ELF entry point, it must point `ram32_start ` (230daa6).

To build:
```
      cargo build \
        --target target.json \
        -Zbuild-std=core \
        -Zbuild-std-features=compiler-builtins-mem
        --features "coreboot"
```
Build `coreboot.rom` with `hypervisor-fw` as a payload.

To run:
```
    qemu-system-x86_64
      -enable-kvm \
      -cpu host \
      -machine type=q35 \
      -drive if=pflash,format=raw,readonly,file=coreboot.rom \
      -drive id=disk,file=bionic-server-cloudimg-amd64-raw.img,if=none \
      -device virtio-blk-pci,drive=disk,disable-legacy=on \
      -nographic \
      -serial mon:stdio \
      -m 8G
```

I confirmed that it works on both Linux loader mode and EFI mode.

[1] https://www.coreboot.org/API

### CODE REVIEW

**Comment by retrage** [comment]
Does anyone have comments?

File: layout.ld [hunk 1]
```diff
@@ -1,4 +1,4 @@
-ENTRY(rust64_start) /* This firmware doesn't use the ELF entrypoint */
+ENTRY(ram32_start) /* This firmware doesn't use the ELF entrypoint */
```
Comment (rbradford) [inline layout.ld:1]: Please update the comment.

File: src/coreboot.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    size: CBUInt64,
+    entry_type: u32,
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct CBUInt64 {
+    lo: u32,
+    hi: u32,
+}
```
Comment (rbradford) [inline src/coreboot.rs:54]: Wow...

**Review by rbradford — COMMENTED** [review]
This looks fine but I think it probably needs an integration test.

File: layout.ld (hunk 1 above)
Comment (retrage) [reply layout.ld]: Updated with commit 80b6071.

File: src/coreboot.rs (hunk 2 above)
Comment (retrage) [reply src/coreboot.rs]: Oops, I replaced the mess with `#[repr(packed, C)]` in 94e8240

**Comment by retrage** [comment]
For integration tests, I would like to add a shell script like `run_integration_tests.sh`.

**Comment by retrage** [comment]
> This looks fine but I think it probably needs an integration test.

I pushed integration tests for coreboot as b7e1ff9. It passes all integration tests.

**Comment by retrage** [comment]
The coreboot integration tests clone coreboot repository, checkout submodules, and build a binutils/gcc cross compiler every time. It's heavy and time-consuming process to just building a BIOS rom. I think we can improve this by providing pre-built packages for CI.

**Review by rbradford — APPROVED** [review]
@josephlr What do you think?

**Comment by retrage** [comment]
@josephlr Any updates?

**Comment by josephlr** [comment]
This looks reasonable to me, and we should be able to reuse some of this for #24.

I haven't had the chance to review this in extreme detail, but if it boots, we can merge this for now and clean it up later.

Long term, we will probably want to use some sort of CI caching (say w/ Github Actions) to speed stuff up.

**Review by josephlr — APPROVED** [review]

**Pull request merged**